File: native/src/event_source.cpp
#include "gvfi/event_source.hpp"
#include "gvfi/work_loop.hpp"

#include <utility>

namespace gvfi {

EventSource::EventSource(std::string name) : name_(std::move(name)) {}

EventSource::~EventSource() = default;

void EventSource::signalWork() {
  work_to_do_.store(true, std::memory_order_release);
  if (loop_) {
    loop_->signalWork();
  }
}

void EventSource::invokeAction() {
  if (action_) {
    action_();
  }
}

bool EventSource::checkForWork() {
  if (!enabled_.load(std::memory_order_acquire)) {
    return false;
  }
  if (!work_to_do_.exchange(false, std::memory_order_acq_rel)) {
    return false;
  }
  invokeAction();
  return work_to_do_.load(std::memory_order_acquire);
}

CommandGate::CommandGate(std::string name) : EventSource(std::move(name)) {}

void CommandGate::submit(Action action) {
  if (!action) {
    return;
  }
  {
    std::lock_guard<std::mutex> lock(mu_);
    queue_.push_back(std::move(action));
  }
  signalWork();
}

bool CommandGate::checkForWork() {
  if (!enabled_.load(std::memory_order_acquire)) {
    return false;
  }
  std::vector<Action> batch;
  {
    std::lock_guard<std::mutex> lock(mu_);
    batch.swap(queue_);
  }
  for (auto& fn : batch) {
    if (fn) {
      fn();
    }
  }
  std::lock_guard<std::mutex> lock(mu_);
  return !queue_.empty();
}

TimerEventSource::TimerEventSource(std::string name)
    : EventSource(std::move(name)) {}

void TimerEventSource::setIntervalMs(std::uint64_t interval_ms) {
  interval_ms_.store(interval_ms, std::memory_order_release);
}

void TimerEventSource::arm(bool repeating) {
  repeating_.store(repeating, std::memory_order_release);
  const auto ms = interval_ms_.load(std::memory_order_acquire);
  {
    std::lock_guard<std::mutex> lock(timing_mu_);
    next_fire_ = std::chrono::steady_clock::now() +
                 std::chrono::milliseconds(ms);
  }
  armed_.store(true, std::memory_order_release);
  signalWork();
}

void TimerEventSource::cancel() {
  armed_.store(false, std::memory_order_release);
}

bool TimerEventSource::checkForWork() {
  if (!enabled_.load(std::memory_order_acquire) ||
      !armed_.load(std::memory_order_acquire)) {
    return false;
  }
  const auto now = std::chrono::steady_clock::now();
  bool fire = false;
  {
    std::lock_guard<std::mutex> lock(timing_mu_);
    if (now >= next_fire_) {
      fire = true;
      if (repeating_.load(std::memory_order_acquire)) {
        next_fire_ =
            now + std::chrono::milliseconds(
                      interval_ms_.load(std::memory_order_acquire));
      } else {
        armed_.store(false, std::memory_order_release);
      }
    }
  }
  if (fire) {
    invokeAction();
  }
  // Keep the loop awake while a timer is armed (cheap poll).
  if (armed_.load(std::memory_order_acquire) && loop_) {
    loop_->signalWork();
  }
  return false;
}

}  // namespace gvfi

File: ECCV2022-RIFE/gvfi_runtime/event_source.py
from __future__ import annotations

import threading
import time
from typing import Callable, List, Optional

Action = Callable[[], None]


class EventSource:
    def __init__(self, name: str = "") -> None:
        self.name = name
        self._enabled = True
        self._work_to_do = False
        self._action: Optional[Action] = None
        self._loop = None
        self._lock = threading.Lock()

    def set_enabled(self, enabled: bool) -> None:
        self._enabled = enabled

    @property
    def enabled(self) -> bool:
        return self._enabled

    def set_action(self, action: Optional[Action]) -> None:
        self._action = action

    def set_work_loop(self, loop) -> None:
        self._loop = loop

    def signal_work(self) -> None:
        with self._lock:
            self._work_to_do = True
        if self._loop is not None:
            self._loop.signal_work()

    def _invoke_action(self) -> None:
        if self._action:
            self._action()

    def check_for_work(self) -> bool:
        if not self._enabled:
            return False
        with self._lock:
            if not self._work_to_do:
                return False
            self._work_to_do = False
        self._invoke_action()
        with self._lock:
            return self._work_to_do


class CommandGate(EventSource):
    def __init__(self, name: str = "command-gate") -> None:
        super().__init__(name)
        self._queue: List[Action] = []
        self._qlock = threading.Lock()

    def submit(self, action: Action) -> None:
        if action is None:
            return
        with self._qlock:
            self._queue.append(action)
        self.signal_work()

    def check_for_work(self) -> bool:
        if not self._enabled:
            return False
        with self._qlock:
            batch = self._queue
            self._queue = []
        for fn in batch:
            fn()
        with self._qlock:
            return bool(self._queue)


class TimerEventSource(EventSource):
    def __init__(self, name: str = "timer") -> None:
        super().__init__(name)
        self._interval_s = 0.0
        self._repeating = False
        self._armed = False
        self._next_fire = 0.0
        self._tlock = threading.Lock()

    def set_interval_ms(self, interval_ms: float) -> None:
        self._interval_s = max(0.0, float(interval_ms) / 1000.0)

    def arm(self, repeating: bool = True) -> None:
        with self._tlock:
            self._repeating = repeating
            self._next_fire = time.monotonic() + self._interval_s
            self._armed = True
        self.signal_work()

    def cancel(self) -> None:
        with self._tlock:
            self._armed = False

    def check_for_work(self) -> bool:
        if not self._enabled or not self._armed:
            return False
        now = time.monotonic()
        fire = False
        with self._tlock:
            if now >= self._next_fire:
                fire = True
                if self._repeating:
                    self._next_fire = now + self._interval_s
                else:
                    self._armed = False
        if fire:
            self._invoke_action()
        if self._armed and self._loop is not None:
            self._loop.signal_work()
        return False
